src/coprocessor/codec/mysql/duration.rs

```diff
@@ -253,97 +254,222 @@ mod parser {
+        self.set_neg(false);
+        self.set_fsp(0);
+        self.to_bits() == 0
+    }
+
+    /// Returns the absolute value of `Duration`
+    pub fn abs(mut self) -> Self {
+        self.set_neg(false);
+        self
+    }
+
+    /// Returns the fractional part of `Duration`, in whole microseconds.
+    pub fn subsec_micros(self) -> u32 {
+        self.micros()
+    }
+
+    /// Returns the number of whole seconds contained by this Duration.
+    pub fn to_secs(self) -> i32 {
+        let secs =
+            (self.hours() * SECS_PER_HOUR + self.minutes() * SECS_PER_MINUTE + self.secs()) as i32;
+
+        if self.get_neg() {
+            -secs
+        } else {
+            secs
+        }
     }
 
-    pub fn to_secs(&self) -> f64 {
-        let res = self.dur.as_secs() as f64 + f64::from(self.dur.subsec_nanos()) * 10e-9;
-        if self.neg {
+    /// Returns the number of seconds contained by this Duration as f64.
+    /// The returned value does include the fractional (nanosecond) part of the duration.
+    pub fn to_secs_f64(self) -> f64 {
+        let res = f64::from(self.to_secs().abs()) + f64::from(self.subsec_micros()) * 1e-6;
+
+        if self.get_neg() {
             -res
         } else {
             res
         }
     }
 
-    pub fn is_zero(&self) -> bool {
-        self.to_nanos() == 0
-    }
+    /// Returns the `Duration` in whole nanoseconds
+    pub fn to_nanos(self) -> i64 {
+        let nanos =
+            i64::from(self.to_secs().abs()) * NANOS_PER_SEC + i64::from(self.micros() * 1000);
 
-    pub fn to_nanos(&self) -> i64 {
-        let nanos = self.dur.as_secs() as i64 * NANOS_PER_SEC + i64::from(self.dur.subsec_nanos());
-        if self.neg {
+        if self.get_neg() {
             -nanos
         } else {
             nanos
         }
     }
 
+    /// Constructs a `Duration` from `nanos` with `fsp`
     pub fn from_nanos(nanos: i64, fsp: i8) -> Result<Duration> {
+        let fsp = check_fsp(fsp)?;
         let neg = nanos < 0;
-        let nanos = nanos.abs();
 
-        let dur = StdDuration::new(
-            (nanos / NANOS_PER_SEC) as u64,
-            (nanos % NANOS_PER_SEC) as u32,
-        );
-        Duration::new(dur, neg, fsp)
+        let nanos = nanos.checked_abs().ok_or(invalid_type!("nanos overflow"))?;
+
+        let mut micros = (nanos % NANOS_PER_SEC / 1000) as u32;
+        let secs = nanos / NANOS_PER_SEC;
+
+        let mut hours = (secs / 3600) as u32;
+        let mut minutes = (secs % 3600 / 60) as u32;
+        let mut secs = (secs % 60) as u32;
+
+        Duration::round(&mut hours, &mut minutes, &mut secs, &mut micros, fsp)?;
+        Ok(Duration::new(neg, hours, minutes, secs, micros, fsp))
     }
 
-    pub fn new(dur: StdDuration, neg: bool, fsp: i8) -> Result<Duration> {
-        check_dur(&dur)?;
-        Ok(Duration {
-            dur,
-            neg,
-            fsp: check_fsp(fsp)?,
-        })
+    pub fn from_millis(millis: i64, fsp: i8) -> Result<Duration> {
```

Comment:
The logic is highly similar to `from_nanos`. How about:

Implement `from_micros`. Then, for `from_nanos`, simply call `Duration::from_micros(nanos / 1000)` and for `from_millis`, simply call `Duration::from_micros(millis * 1000)`.